File: clean-realtime-caption-translator/apps/desktop-shell/src/services/diagnostics/secretRedactor.ts
const SECRET_KEY_PATTERN = /(authorization|bearer|api[_-]?key|dashscope_api_key|game_tts_api_key|token|secret|password)/i;
const BEARER_PATTERN = /Bearer\s+[A-Za-z0-9._~+/=-]+/gi;
const AUTH_LINE_PATTERN = /(Authorization\s*[:=]\s*)([^\s,;}]+)/gi;

export type Redacted<T> = T;

export function redactSecrets<T>(value: T): Redacted<T> {
  if (typeof value === 'string') {
    return value.replace(BEARER_PATTERN, 'Bearer [REDACTED]').replace(AUTH_LINE_PATTERN, '$1[REDACTED]') as Redacted<T>;
  }
  if (Array.isArray(value)) {
    return value.map((item) => redactSecrets(item)) as Redacted<T>;
  }
  if (value && typeof value === 'object') {
    const result: Record<string, unknown> = {};
    for (const [key, item] of Object.entries(value)) {
      result[key] = SECRET_KEY_PATTERN.test(key) ? '[REDACTED]' : redactSecrets(item);
    }
    return result as Redacted<T>;
  }
  return value;
}

File: clean-realtime-caption-translator/apps/desktop-shell/src-tauri/src/secret_redactor.rs
use serde_json::Value;

pub fn redact_text(input: &str) -> String {
    let mut output = input.to_string();
    for marker in ["Authorization", "Bearer", "DASHSCOPE_API_KEY", "GAME_TTS_API_KEY"] {
        if output.to_lowercase().contains(&marker.to_lowercase()) {
            output = "[REDACTED]".to_string();
        }
    }
    output
}

pub fn redact_json(value: Value) -> Value {
    match value {
        Value::Object(map) => Value::Object(
            map.into_iter()
                .map(|(key, item)| {
                    let lower = key.to_lowercase();
                    if lower.contains("key")
                        || lower.contains("token")
                        || lower.contains("secret")
                        || lower.contains("password")
                        || lower == "authorization"
                    {
                        (key, Value::String("[REDACTED]".to_string()))
                    } else {
                        (key, redact_json(item))
                    }
                })
                .collect(),
        ),
        Value::Array(items) => Value::Array(items.into_iter().map(redact_json).collect()),
        Value::String(text) => Value::String(redact_text(&text)),
        other => other,
    }
}
